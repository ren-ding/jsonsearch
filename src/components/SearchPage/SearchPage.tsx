import React, { Component } from 'react';
import ZendeskSearch, {ZendeskSearchInterface} from '../../api/ZendeskSearch';
import OrganizationContainer from '../../model/OrganizationContainer';
import UserContainer from '../../model/UserContainer';
import TicketContainer from '../../model/TicketContainer';
import SearchFilter from '../SearchFilter';
import {Option} from 'react-dropdown';
import './style/search-page.css';

interface SearchPagePropsInterface {
}

interface SearchPageStatesInterface {
    zendeskSearch: ZendeskSearch,
    searchResults: Array<OrganizationContainer|UserContainer|TicketContainer>,
    selectedSearchTable: SearchTable,
    searchableFields: string,
    searchValue: string
}

export enum SearchTable {
    Organizations,
    Users,
    Tickets
}

/// <summary>
/// SearchPage component is a container component using for searching data
/// The component contains: a header component, a filter component, a result display component and a search button
/// </summary>
export default class SearchPage extends Component<SearchPagePropsInterface, SearchPageStatesInterface>
                                implements ZendeskSearchInterface {
    constructor(props:SearchPagePropsInterface) {
        super(props);
        this.state = {
          zendeskSearch: new ZendeskSearch(),
          searchResults:[],
          selectedSearchTable: SearchTable.Organizations,
          searchableFields: "",
          searchValue: ""
        };
    }

    render() {
        const ddlSearchTableOptions = [
            {value: SearchTable.Organizations.toString(), label:'Organizations'},
            {value: SearchTable.Users.toString(), label:'Users'},
            {value: SearchTable.Tickets.toString(), label:'Tickets'}
        ];

        return (
        <div className='search-page-wrapper'>
            <ZendeskSearchHeader/>
            <SearchFilter
             searchTableDropdownList = { ddlSearchTableOptions }
             onChange = {this.onSearchTableDropdownListChange}
             selectedSearchTable = {this.state.selectedSearchTable}
             onSearchableFieldsChange = {this.onSearchableFieldsChange}
             onSearchValueChange = {this.onSearchValueChange}
            />
            <button type="button" className='ui blue button search-button' onClick={this.onSearchButtonClick}>Search</button>
        </div>);
    }

    searchOrganizationWithRelatedInfo(searchableFields:Array<string>, searchValue: string) {
        return this.state.zendeskSearch.searchOrganizationWithRelatedInfo(searchableFields, searchValue);
    } 

    searchUserWithRelatedInfo(searchableFields:Array<string>, searchValue: string) {
        return this.state.zendeskSearch.searchUserWithRelatedInfo(searchableFields, searchValue);;
    }

    searchTicketWithRelatedInfo(searchableFields:Array<string>, searchValue: string) {
        return this.state.zendeskSearch.searchTicketWithRelatedInfo(searchableFields, searchValue);;
    }

    onSearchButtonClick = () => {
        const searchableFieldsArray = this.state.searchableFields.split(',');

        if(this.state.selectedSearchTable === null || searchableFieldsArray.length === 0) return;
        switch(this.state.selectedSearchTable){
            case SearchTable.Organizations:
                this.setState({
                    searchResults: this.searchOrganizationWithRelatedInfo(searchableFieldsArray, this.state.searchValue)
                });
                break;
            case SearchTable.Users:
                this.setState({
                    searchResults: this.searchUserWithRelatedInfo(searchableFieldsArray, this.state.searchValue)
                });
                break;
            case SearchTable.Tickets:
                this.setState({
                    searchResults: this.searchTicketWithRelatedInfo(searchableFieldsArray, this.state.searchValue)
                });
                break;
            default:
                return;
        }
    }

    onSearchTableDropdownListChange = (option:Option) => {
        this.setState({
            selectedSearchTable: parseInt(option.value)
        })
    }

    onSearchableFieldsChange = (event: {target: { value: string }}) => {
        this.setState({
            searchableFields:event.target.value
        });
    }

    onSearchValueChange = (event: {target: { value: string }}) => {
        this.setState({
            searchValue:event.target.value
        });
    }

}

const ZendeskSearchHeader = () => {
    //unfortunitely, keys of interface in typescript still not supported
    //https://github.com/Microsoft/TypeScript/issues/13267
    const keysOfOrganization = [
        '_id','url','external_id','name','domain_names',
        'created_at','details','shared_tickets','tags'
    ];

    const keysOfUser = [
        '_id','url','external_id','name','alias','created_at','active','verified',
        'shared','locale','timezone','last_login_at','email','phone','signature',
        'organization_id','tags','suspended','role'
    ];
    
    const keysOfTicket = [
            '_id','url','external_id','created_at','type','subject','description',
            'priority','status','submitter_id','assignee_id','organization_id',
            'tags','has_incidents','due_at','via'
    ];
    
    return (
    <div className='zendesk-search-header'>
        <h1 className="ui center aligned header">Zendesk Search</h1>
        <h2 className="ui header">
            <i className="search icon"></i>
            <div className="content">Searchable Fields</div>
        </h2>
        <table className="ui celled striped table">
            <thead>
                <tr>
                    <th colSpan={2}>
                        Searchable Fields
                    </th>
                </tr>
            </thead>
            <tbody>
                <tr>
                    <td>User Fields</td>
                    <td>{keysOfUser.join(',')}</td>
                </tr>
                <tr>
                    <td>Organization Fields</td>
                    <td>{keysOfOrganization.join(',')}</td>
                </tr>
                <tr>
                    <td>Ticket Fields</td>
                    <td>{keysOfTicket.join(',')}</td>
                </tr>
            </tbody>
        </table>
    </div>);
};
