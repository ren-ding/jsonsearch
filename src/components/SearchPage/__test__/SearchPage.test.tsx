import React from 'react';
import SearchPage from '../SearchPage';
import renderer from 'react-test-renderer';

describe('SearchPage', () => {
    it('should render to match snapshot', () => {
        const tree = renderer.create(<SearchPage />).toJSON();
        expect(tree).toMatchSnapshot();
      });

});