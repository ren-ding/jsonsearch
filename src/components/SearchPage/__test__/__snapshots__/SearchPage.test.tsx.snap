// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage should render to match snapshot 1`] = `
<div>
  <div
    className="zendesk-search-header"
  >
    Zendesk Search
    <p>
      Searchable fields
    </p>
    <p>
      Organization table:
    </p>
    _id, 
    url, 
    external_id, 
    name, 
    domain_names, 
    created_at, 
    details, 
    shared_tickets, 
    tags, 
    <p>
      User table:
    </p>
    _id, 
    url, 
    external_id, 
    name, 
    alias, 
    created_at, 
    active, 
    verified, 
    shared, 
    locale, 
    timezone, 
    last_login_at, 
    email, 
    phone, 
    signature, 
    organization_id, 
    tags, 
    suspended, 
    role, 
    <p>
      Ticket table:
    </p>
    _id, 
    url, 
    external_id, 
    created_at, 
    type, 
    subject, 
    description, 
    priority, 
    status, 
    submitter_id, 
    assignee_id, 
    organization_id, 
    tags, 
    has_incidents, 
    due_at, 
    via, 
  </div>
  <div>
    <div
      className="Dropdown-root"
    >
      <div
        aria-haspopup="listbox"
        className="Dropdown-control"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <div
          className="Dropdown-placeholder is-selected"
        >
          Organizations
        </div>
        <div
          className="Dropdown-arrow-wrapper"
        >
          <span
            className="Dropdown-arrow"
          />
        </div>
      </div>
    </div>
    <div>
      <label>
        SearchableFields(e.g. name,tags): 
      </label>
      <input
        onChange={[Function]}
        type="text"
      />
    </div>
    <div>
      <label>
        Search Value: 
      </label>
      <input
        onChange={[Function]}
        type="text"
      />
    </div>
  </div>
  <button
    className="search-button"
    onClick={[Function]}
    type="button"
  >
    Search
  </button>
</div>
`;
